src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -1081,12 +1110,41 @@ public IEnumerable<KeyValuePair<string, string>> EnumerateMetadata()
             /// Equivalent to calling <see cref="SetMetadata(string,string)"/> for each item in <paramref name="metadata"/>.
             /// </summary>
             /// <param name="metadata">The metadata to set.</param>
-            public void ImportMetadata(IEnumerable<KeyValuePair<string, string>> metadata)
+            public void ImportMetadata(IEnumerable<KeyValuePair<string, string>> metadata) =>
+                ImportMetadata(metadata, validateKeys: true);
+
+            /// <summary>
+            /// Sets the given metadata.
+            /// </summary>
+            /// <param name="metadata">The metadata to set.</param>
+            /// <param name="validateKeys">If true, all imported metadata keys will be validated for reserved names.</param>
+            /// <remarks>
+            /// If importing from another ProjectItemInstance.TaskItem instance, the caller should not set validateKeys
+            /// since all incoming metadata will already have been validated.
+            /// </remarks>
+            private void ImportMetadata(IEnumerable<KeyValuePair<string, string>> metadata, bool validateKeys)
             {
                 ProjectInstance.VerifyThrowNotImmutable(_isImmutable);
 
-                _directMetadata ??= new CopyOnWritePropertyDictionary<ProjectMetadataInstance>();
-                _directMetadata.ImportProperties(metadata.Select(kvp => new ProjectMetadataInstance(kvp.Key, kvp.Value, allowItemSpecModifiers: true)));
+                if (DirectMetadataCount == 0 && metadata.GetType() == typeof(ImmutableDictionary<string, string>))
```

Comment:
Is there a specific reason to type check via `GetType()` rather than with is/as?